File: study1/week3/evaluate-the-bracket-pairs-of-a-string.java
// https://leetcode.com/problems/evaluate-the-bracket-pairs-of-a-string
// Time Complexity: O(n)
// Space Complexity: O(n), because of hashmap

class Solution {
    public String evaluate(String s, List<List<String>> knowledge) {
        StringBuilder result = new StringBuilder();
        HashMap<String, String> wordMap = new HashMap<>();  // key: word to be changed, value: changing word

        for (List<String> word : knowledge) {
            wordMap.put(word.get(0), word.get(1));
        }

        for (int i = 0; i < s.length(); ++i) {
            char ch = s.charAt(i);
            if (ch == '(') {    // bracket found
                int end = s.indexOf(")", i);
                result.append(wordMap.getOrDefault(s.substring(i+1, end), "?"));
                i = end;  // change index to the end
            } else {
                result.append(ch);
            }
        }

        return result.toString();
    }
}

File: study1/solutions/week3/evaluate_the_bracket_pairs_of_a_string.py
# https://leetcode.com/problems/evaluate-the-bracket-pairs-of-a-string/description/
class Solution:
    def evaluate(self, s: str, knowledge: List[List[str]]) -> str:
        knowledge_mapping = dict(knowledge)

        words = []
        start = 0
        for i, char in enumerate(s):
            if char == '(':
                words.append(s[start:i])
                start = i + 1
                continue

            if char == ')':
                words.append(knowledge_mapping.get(s[start:i], '?'))
                start = i + 1
                continue
            

        words.append(s[start:i + 1])

        return "".join(words)